### PR TITLE
[TAGS] Change ES mapping and API endpoints

### PR DESCRIPTION
**Pull request by simonmeyerrr**
/!\ Elastic search datas have to be deleted and reingested

### Change ES mapping for tags using Nested
 
Exemple in ES:
```
"tags": [
	{
		"key": "a_tag_key",
		"tag": "prod"
	},
	{
		"key": "another_key",
		"tag": "other_tag"
	}, ...
]
```

### Change  endpoints usage and response

- ### `/costs/tags/keys`
Require to be authentified
Query args:
  - aa: (Optional) list of aws accounts to filter separated by a comma, default: all
  - begin: date for time range
  - end: date for time range

Exemple:
```
$> curl $API/costs/tags/keys?aa=$ACCOUNT_ID&begin=êBEGIN_DATE&end=$END_DATE
[
  "tagkey",
  "another_tagkey"
]
```

- ### `/costs/tags/values`
Require to be authentified
Query args:
  - aa: (Optional) list of aws accounts to filter separated by a comma, default: all
  - begin: date for time range
  - end: date for time range
  - keys: (Optional) list of keys to filter separated by a comma, default: all
  - by: filter (product / region / account / availabilityzone)

Exemple:
```
$> curl $API/costs/tags/values?aa=$ACCOUNT_ID&begin=$BEGIN_DATE&end=$END_DATE&keys=tagkey,another_tagkey&by=product

{
	"tagkey": [
		{
			"tag": "prod",
			"costs": [
				{
					"item": "AmazonS3",
					"cost": 64.865746
				},
				{
					"item": "AmazonEC2",
					"cost": 159.64454746
				}, ...
			]
		},
		{
			"tag": "staging",
			"costs": [
				{
					"item": "AmazonS3",
					"cost": 4.8658786746
				},
				{
					"item": "AmazonEC2",
					"cost": 19.64876746
				}, ...
			]
		}, ...
	],
	"another_tagkey": [
		...
	]
}
```

### CODE REVIEW

File: costs/tags/tags_values.go [hunk 1]
```diff
@@ -31,55 +31,70 @@ import (
...
+									} `json:"cost"`
+								} `json:"buckets"`
+							} `json:"filter"`
+						} `json:"rev"`
+					} `json:"buckets"`
+				} `json:"tags"`
 			} `json:"buckets"`
-		}
+		} `json:"keys"`
 	}
```
Comment (thibautcornolti) [inline costs/tags/tags_values.go]: Please split your structure.

File: costs/tags/tags_values.go [hunk 2]
```diff
@@ -31,55 +31,70 @@ import (
...
-
-	// result format of the endpoint
-	TagsValuesResponse []TagsValues
 )
 
 // getTagsValuesWithParsedParams will parse the data from ElasticSearch and return it
-func getTagsValuesWithParsedParams(ctx context.Context, params tagsValuesQueryParams, user users.User) (int, interface{}){
-	var response = TagsValuesResponse{}
+func getTagsValuesWithParsedParams(ctx context.Context, params tagsValuesQueryParams, user users.User) (int, interface{}) {
+	response := make(map[string][]TagsValues)
```
Comment (thibautcornolti) [inline costs/tags/tags_values.go]:
```
type TagsValuesResponse map[string][]TagsValues]
var response TagsValuesResponse
```

File: costs/tags/tags_values.go [hunk 3]
```diff
@@ -128,3 +149,29 @@ func createQueryAccountFilter(accountList []string) *elastic.TermsQuery {
...
+
+// arrayContainsString returns true if a string is present in an array of string
+func arrayContainsString(s []string, e string) bool {
+	for _, a := range s {
+		if a == e {
+			return true
+		}
+	}
+	return false
+}
```
Comment (thibautcornolti) [inline costs/tags/tags_values.go]: New line

File: costs/tags/tags_values.go [hunk 4]
```diff
@@ -128,3 +149,29 @@ func createQueryAccountFilter(accountList []string) *elastic.TermsQuery {
...
+	case "product":
+		return "productCode"
+	case "region":
+		return "region"
+	case "account":
+		return "usageAccountId"
+	case "availabilityzone":
+		return "availabilityZone"
+	default:
+		return "error"
```
Comment (thibautcornolti) [inline costs/tags/tags_values.go]: Why didn't you use an hard-coded array?

**Review by thibautcornolti — DISMISSED** [review]

**Review by zanchi-r — APPROVED** [review]

**Pull request merged**